src/main/scala/it/pagopa/interop/catalogprocess/api/impl/ProcessApiServiceImpl.scala

```diff
@@ -156,32 +164,52 @@ final case class ProcessApiServiceImpl(
   /** Code: 200, Message: A list of E-Service, DataType: Seq[EService]
     * Code: 500, Message: Internal Server Error, DataType: Problem
     */
-  override def getEServices(
-    producerId: Option[String],
-    consumerId: Option[String],
-    agreementState: String,
-    status: Option[String]
-  )(implicit
+  override def getEServices(nameFilter: Option[String], producersId: String, states: String, offset: Int, limit: Int)(
+    implicit
     contexts: Seq[(String, String)],
     toEntityMarshallerEServicearray: ToEntityMarshaller[Seq[EService]]
   ): Route = authorize(ADMIN_ROLE, API_ROLE, SECURITY_ROLE, M2M_ROLE) {
-    logger.info("Getting e-service with producer = {}, consumer = {} and state = {}", producerId, consumerId, status)
+    logger.info(s"Getting e-service with producers = $producersId, states = $states")
+
+    // TODO Improve code
+    def paramsToQuery(): Either[Throwable, Bson] = for {
+      apiStatesFilters <- parseArrayParameters(states).traverse(EServiceDescriptorState.fromValue)
+      persistenceStatesFilter = apiStatesFilters
+        .map(convertFromApiDescriptorState)
+        .map(_.toString)
+        .map(Filters.eq("data.descriptors.state", _))
+      statesQuery = if (persistenceStatesFilter.isEmpty) None else Filters.or(persistenceStatesFilter: _*).some
+
+      nameQuery = nameFilter.map(Filters.regex("data.name", _, "i"))
+
+      producerIdFilters = parseArrayParameters(producersId).map(Filters.eq("data.producerId", _))
+      producerIdQuery   = if (producerIdFilters.isEmpty) None else Filters.or(producerIdFilters: _*).some
+      filters           = nameQuery.toList ++ producerIdQuery.toList ++ statesQuery.toList
+    } yield if (filters.isEmpty) Filters.empty() else Filters.and(filters: _*)
+
```

Comment:
the `if (list.isEmpty) None else Filter.and/or(list:_*)` maybe can be extracted in a method.

Plus I will move the `producerIdFilters = parseArrayParameters(producersId).map(Filters.eq("data.producerId", _))` in order to parse everything as first step and then compile to Filters

Reply:
yup this part will be completely refactored, stay tuned 📻